rockpi4.mk

```diff
@@ -0,0 +1,231 @@
+linux-defconfig: $(LINUX_PATH)/.config
+
+LINUX_COMMON_FLAGS += ARCH=arm64 Image rockchip/rk3399-rock-pi-4b.dtb
+
+.PHONY: linux
+linux: linux-common
+
+$(LINUX_PATH)/arch/arm64/boot/Image.gz: linux
+	gzip -c $(LINUX_PATH)/arch/arm64/boot/Image >$@
+
+.PHONY: linux-defconfig-clean
+linux-defconfig-clean: linux-defconfig-clean-common
+
+LINUX_CLEAN_COMMON_FLAGS += ARCH=arm64
+
+.PHONY: linux-clean
+linux-clean: linux-clean-common
+
+LINUX_CLEANER_COMMON_FLAGS += ARCH=arm64
+
+.PHONY: linux-cleaner
+linux-cleaner: linux-cleaner-common
+
+################################################################################
+# OP-TEE
+################################################################################
+OPTEE_OS_PLATFORM = rockchip-rk3399
+OPTEE_OS_COMMON_FLAGS += CFG_ENABLE_EMBEDDED_TESTS=y
+
+.PHONY: optee-os
+optee-os: optee-os-common
+
+.PHONY: optee-os-clean
+optee-os-clean: optee-os-clean-common
+
+clean: optee-os-clean
+
+################################################################################
+# Boot image, shall be copied to SD card
+################################################################################
+# U-Boot offset comes from CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR=0x4000
+# Partition no. 5 ends at 12288 + BR2_TARGET_ROOTFS_EXT2_SIZE (in kiB)
+.PHONY: boot-img
+boot-img: u-boot buildroot $(LINUX_PATH)/arch/arm64/boot/Image.gz
+	mkdir -p $(BINARIES_PATH)
+	rm -f $(BOOT_IMG)
+	truncate -s 128MiB $(BOOT_IMG)
+	parted -s $(BOOT_IMG) \
+		unit kiB \
+		mklabel gpt \
+		mkpart idbloader 32 4032 \
+		mkpart primary fat32 4032 4096 \
+		mkpart primary fat32 4096 8192 \
+		mkpart uboot 8192 12288 \
+		mkpart root fat32 12288 126976
+	sgdisk -u 5:17d61bff-8fdc-4089-b675-9be21b9f6ac7 $(BOOT_IMG)
+	dd if=$(UBOOT_PATH)/idbloader.img of=$(BOOT_IMG) bs=1kiB seek=32 conv=notrunc
+	dd if=$(UBOOT_PATH)/u-boot.itb of=$(BOOT_IMG) bs=1kiB seek=8192 conv=notrunc
+	e2mkdir $(ROOT_IMG):/boot
+	e2cp $(LINUX_PATH)/arch/arm64/boot/Image.gz $(ROOT_IMG):/boot
+	e2cp $(LINUX_PATH)/arch/arm64/boot/dts/rockchip/rk3399-rock-pi-4b.dtb $(ROOT_IMG):/boot
+	dd if=$(ROOT_IMG) of=$(BOOT_IMG) bs=1kiB seek=12288 conv=notrunc
+
+.PHONY: boot-img-clean
+boot-img-clean:
+	rm -f $(BOOT_IMG)
+
+clean: boot-img-clean
+
+################################################################################
+# rkdeveloptool
+################################################################################
+
+$(RKDEVELOPTOOL_PATH)/Makefile:
+	cd $(RKDEVELOPTOOL_PATH) && \
+		autoreconf -i && \
+		./configure CXXFLAGS=-Wno-format-truncation
+
+$(RKDEVELOPTOOL_BIN): $(RKDEVELOPTOOL_PATH)/Makefile
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH)
+
+rkdeveloptool: $(RKDEVELOPTOOL_BIN)
+
+rkdeveloptool-clean:
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH) clean
+
+rkdeveloptool-distclean:
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH) clean
+
+clean: rkdeveloptool-clean
+
+$(LOADER_BIN):
+	cd $(BINARIES_PATH) && \
+		wget https://dl.radxa.com/rockpi/images/loader/$(notdir $(LOADER_BIN))
```

Comment:
We should probably remove the downloaded file if this fails for some reason, network timeout in the middle for instance.

Reply:
I agree. I will add `.DELETE_ON_ERROR` to take care of that.